src/reference.rs

```diff
@@ -95,21 +97,43 @@ impl Reference {
     pub fn with_digest(registry: String, repository: String, digest: String) -> Self {
         Self {
             registry,
+            mirror_registry: None,
             repository,
             tag: None,
             digest: Some(digest),
         }
     }
 
+    /// Clone the Reference for the same image with a new digest.
+    pub fn clone_with_digest(&self, digest: String) -> Self {
+        Self {
+            registry: self.registry.clone(),
+            mirror_registry: self.mirror_registry.clone(),
+            repository: self.repository.clone(),
+            tag: None,
+            digest: Some(digest),
+        }
+    }
+
+    /// Set a pull mirror registry for this reference.
+    ///
+    /// The mirror registry will be used to resolve the image, the original registry
+    /// is available via the [`Reference::ns`] function.
+    pub fn set_mirror_registry(&mut self, registry: String) {
```

Comment:
We should probably have this and the `ns` function be `#[doc(hidden)]` with information in their doc strings about this not being stable yet (and as such, is exempt from semver backwards compat guarantees). I am perfectly fine with it being in here if things like containerd are already using it, but I want to make it clear that this isn't fully supported yet. The other option would be to have something like a `ClientExp` trait that the client could implement for experimental features, but that also feels a bit extra. 

Like you mentioned, it looks like it will land, but things can take a while in the OCI spec, so I want to be extra careful here

Reply:
Sure, I added the `#[doc(hidden)]` and added a note about it not being stable yet, and also added a note that the `ns` query parameter might not work with all registries yet.